### ISSUE TITLE
Issue 002: time-axis log_softmax score と 2D DP component tracer を実装する

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#351]
# Issue 002: time-axis log_softmax score と 2D DP component tracer を実装する

## 背景

GPT の正式入力は `log_softmax score` に固定する。coarse model から出る raw tensor は logits だが、GPT / DP に渡す score は time axis で正規化した log probability とする。

```text
coarse logits
  -> log_softmax(axis=time)
  -> GPT score
```

DP は path 上の score 総和を最大化する。log probability の総和を最大化することで、smoothness constraint 付きの最大尤度 path に近い意味になる。

## 目的

`path_tracing` package に以下を実装する。

1. stable `log_softmax` helper
2. 1 component 用の 2D DP accumulation / backtracking

segment 分割や invalid row handling は Issue 003 で扱う。この issue は「連続した有効 row だけから成る 2D score matrix」を対象にする。

## 非目的

この issue では以下は行わない。

- segment/gap 分割 dispatcher
- coarse/infer.py への接続
- channel tiling merge
- 3D GPT 実装
- raw waveform / pilot / amplitude-ratio GPT

## 変更対象

```text
packages/seisai-engine/src/seisai_engine/pipelines/fbpick/path_tracing/score.py
packages/seisai-engine/src/seisai_engine/pipelines/fbpick/path_tracing/dp2d.py
packages/seisai-engine/src/seisai_engine/pipelines/fbpick/path_tracing/__init__.py
packages/seisai-engine/tests/test_fbpick_path_tracing_dp2d.py
```

### 1. score helper

`score.py` に追加する。

```python
def log_softmax_time(logits: np.ndarray, *, time_axis: int = -1) -> np.ndarray:
    ...
```

要件:

- 入力は floating ndarray。
- finite であること。
- stable 実装にする。

```python
shifted = logits - logits.max(axis=time_axis, keepdims=True)
log_denom = np.log(np.exp(shifted).sum(axis=time_axis, keepdims=True))
score = shifted - log_denom
```

- 戻り値 dtype は `np.float32`。
- time axis の `exp(score)` の和が概ね 1 になる。

```python
def prob_from_log_score(score: np.ndarray) -> np.ndarray:
    ...
```

要件:

- `np.exp(score)` を `float32` で返す。
- finite であること。
- `[0, 1]` 範囲を小さい roundoff は clip してよい。

### 2. 2D component DP

`dp2d.py` に追加する。

```python
@dataclass(frozen=True)
class DP2DComponentResult:
    path_j: np.ndarray
    path_score: np.ndarray
    path_prob: np.ndarray
    accumulated_score: np.ndarray
```

```python
def trace_dp2d_component(
    score_hw: np.ndarray,
    *,
    prob_hw: np.ndarray | None = None,
    max_jump: int,
) -> DP2DComponentResult:
    ...
```

入力:

```text
score_hw.shape = (n_rows, n_time)
prob_hw.shape = score_hw.shape or None
max_jump >= 0
```

validation:

- `score_hw` は 2D floating ndarray。
- `score_hw` は finite。
- `n_rows > 0`, `n_time > 0`。
- `max_jump` は bool ではない int, `>= 0`。
- `prob_hw` がある場合、shape 一致、finite、`[0, 1]`。

DP:

```text
S[0, k] = score[0, k]

S[i, k] = score[i, k] + max S[i - 1, q]
          where q in [k - L, k + L]

p[last] = argmax_k S[last, k]

p[i - 1] = argmax_q S[i - 1, q]
           where q in [p[i] - L, p[i] + L]
```

`L = max_jump`。

実装要件:

- `backpointer` は `int32` で持つ。
- tie break は deterministic にする。`np.argmax` の挙動を利用し、同点では最小 index を選ぶ。
- 境界外の q は使わない。
- `max_jump == 0` は vertical path only になる。
- 戻り値:
  - `path_j`: `int64`, shape `(n_rows,)`
  - `path_score`: `float32`, `score_hw[np.arange(n_rows), path_j]`
  - `path_prob`: `float32`; `prob_hw` がある場合はその selected value、ない場合は `exp(path_score)`。
  - `accumulated_score`: `float32`, shape `(n_rows, n_time)`

計算量:

- MVP は `O(n_rows * n_time * (2L+1))` でよい。
- ただし実装は読みやすさを優先し、`n_rows=256`, `n_time=2048`, `L<=32` 程度で現実的な速度にする。
- 後から sliding-window max に置き換えられるよう、関数境界は保つ。

## テスト要件

新規:

```text
packages/seisai-engine/tests/test_fbpick_path_tracing_dp2d.py
```

最低限:

```python
def test_log_softmax_time_normalizes_each_row() -> None: ...
def test_trace_dp2d_component_follows_clean_ridge() -> None: ...
def test_trace_dp2d_component_suppresses_noisy_jump_when_jump_limited() -> None: ...
def test_trace_dp2d_component_allows_jump_when_max_jump_large() -> None: ...
def test_trace_dp2d_component_max_jump_zero_vertical_path() -> None: ...
def test_trace_dp2d_component_rejects_nonfinite_score() -> None: ...
def test_trace_dp2d_component_selected_prob_comes_from_prob_hw() -> None: ...
```

Synthetic ridge 例:

```python
score = -10 * np.ones((20, 64), dtype=np.float32)
path = np.linspace(10, 20, 20).round().astype(int)
score[np.arange(20), path] = 0.0
```

noisy jump test では、ある row にだけ遠い high score を置き、`max_jump` が小さいと拾わず、大きいと拾えることを確認する。

## 受け入れ条件

- GPT の score input は `log_softmax` に固定される。
- `trace_dp2d_component()` は accumulation / backtracking により constrained global path を返す。
- invalid / segment handling なしの component 単位で deterministic に動く。
- 既存 pipeline はまだ変更されない。

## 推奨確認コマンド

```bash
pytest packages/seisai-engine/tests/test_fbpick_path_tracing_dp2d.py -q
python -m compileall packages/seisai-engine/src/seisai_engine/pipelines/fbpick/path_tracing
```
